worker/uniter/uniter.go

```diff
@@ -672,21 +672,27 @@ func (u *Uniter) init(unitTag names.UnitTag) (err error) {
 	default:
 		return errors.Errorf("unknown model type %q", u.modelType)
 	}
+
 	u.unit, err = u.st.Unit(unitTag)
 	if err != nil {
-		return err
+		if errors.IsNotFound(err) {
+			return u.stopUnitError()
+		}
+		return errors.Trace(err)
 	}
 	if u.unit.Life() == life.Dead {
-		// If we started up already dead, we should not progress further. If we
-		// become Dead immediately after starting up, we may well complete any
-		// operations in progress before detecting it; but that race is fundamental
-		// and inescapable, whereas this one is not.
+		// If we started up already dead, we should not progress further.
+		// If we become Dead immediately after starting up, we may well
+		// complete any operations in progress before detecting it,
+		// but that race is fundamental and inescapable,
+		// whereas this one is not.
```

Comment:
This comment is applicable to the not found case also I think.
Would it be nice to do this?
(also avoids u.stopUnitError() twice)

```
	u.unit, err = u.st.Unit(unitTag)
	if err != nil && !errors.IsNotFound(err) {
		return errors.Trace(err)
    }
	if err != nil || u.unit.Life() == life.Dead {
		// If we started up already dead, we should not progress further. If we
		// If we started up already dead, we should not progress further.
		// become Dead immediately after starting up, we may well complete any
		// If we become Dead immediately after starting up, we may well
		// operations in progress before detecting it; but that race is fundamental
		// complete any operations in progress before detecting it,
		// and inescapable, whereas this one is not.
		// but that race is fundamental and inescapable,
		// whereas this one is not.
        return u.stopUnitError()
    }
```

Reply:
I've moved the comment higher, but I think assessing the error in a single block reads better.